src/quicktype-core/language/Python.ts

```diff
@@ -579,6 +587,24 @@ function makeValue(vol: ValueOrLambda): Sourcelike {
     return vol.value;
 }
 
+function threeToTwo(input: Sourcelike): Sourcelike {
```

Comment:
This seems a weird way of doing things.  What does this function do?  Why don't we generate the `input` in the right way in the first place?  You're basically building a post-processor for quicktype output within quicktype.

Reply:
@schani I feel like the current code is written at the wrong level of abstraction to allow that. You'd really want to work at the Decorated Abstract Syntax Tree level, with the emit/codegen step doing the Python version-specific grammar formation.

Here, I was looking more at what's the fewest lines of code to edit in order to facilitate Python 2.7.

29 lines isn't bad 🖌

Reply:
We don't build syntax trees for output languages.  Look at `withTypeHint` for a start on where to emit the correct code in the first place.

Reply:
I understand that you don't currently, my opinion is that if you did it would simplify the Python 2/3 emission.

Anyway, from my knowledge of the comparison, all you need to do is convert the Python 3.7 type into a [mypy] type string, and you're good to go. That's what I've done, with a TODO to do the same for function signatures.

Reply:
`threeToTwo()` is only used for function/method definitions, right?  Why not instead have a method that takes the name, argument names, their types, and the return type, and emits the function definition line - with type annotations if the options require them - and an optional type annotation comment line if Python 2.7 with typing is selected?  Call it `emitFunction` or `emitDef`, and it takes a function that emits its body, like `emitBlock`.

Reply:
What I meant was that:

```python
def hello() -> None:
    print 'hello'
```

Becomes:
```python
def hello(): # type: () -> None
    print 'hello'
```

So the type annotation isn't actually changing, just `# type:` is appended and the type annotation is moved outside the code and [appended] into the comment.

Clearly this means it's not a different typing, it's just appending what is already there as a comment.

Reply:
I understand.  I don't want code that emits Python 3 and then other code that rewrites it to Python 2.  That is brittle (prone to breakage) and obscures what's going on.  Instead, have an `emitDef` which emits either one, or the other, depending on which option is selected.